python_modules/dagster/dagster/_core/definitions/decorators/asset_decorator.py

```diff
@@ -977,12 +1010,20 @@ def build_asset_outs(asset_outs: Mapping[str, AssetOut]) -> Mapping[AssetKey, Tu
     return outs_by_asset_key
 
 
-def _make_asset_keys(deps: Optional[Union[Set[AssetKey], Set[str]]]) -> Optional[Set[AssetKey]]:
-    """Convert all str items to AssetKey in the set."""
+def _make_asset_keys(
+    deps: Optional[Sequence[Union[CoercibleToAssetKey, AssetsDefinition]]]
+) -> Optional[Set[AssetKey]]:
+    """Convert all items to AssetKey in a set. By putting all of the AsestKeys in a set, it will also deduplicate them.
+    """
     if deps is None:
         return deps
 
-    deps_asset_keys = {
-        AssetKey.from_user_string(dep) if isinstance(dep, str) else dep for dep in deps
-    }
+    deps_asset_keys: Set[AssetKey] = set()
+    for dep in deps:
+        if isinstance(dep, AssetsDefinition):
+            # this will error if the AssetsDefinition is a multi_asset
+            deps_asset_keys.add(dep.key)
```

Comment:
If a user tries to pass a multi asset AssetsDefinition, they will get an error like this
```
dagster._check.CheckError: Invariant failed. Description: Tried to retrieve asset key from an assets definition with multiple asset keys: ["asset_1"], ["asset_2"]
```

This doesn't give them much direction about what to fix - maybe we should catch and reraise a more useful message?

Reply:
We detect the error right at the entry point at throw immediately. Always good to catch errors as close to the user code as possile.